omniscidb/QueryOptimizer/CanonizeQuery.cpp

```diff
@@ -351,10 +351,127 @@ void expandCompoundAggregates(QueryDag& dag) {
+    // we currently only set for distributed, but could also be set to support
+    // multi-frag window function inputs, either if we can detect that an input table
+    // is multi-frag up front, or using a retry mechanism like we do for join filter
+    // push down.
+    // TODO(todd): Investigate a viable approach for the above.
+    // 2. Regardless of #1, if the window function project node is preceded by a
+    // filter node. This is required both for correctness and to avoid pulling
+    // all source input columns into memory since non-coalesced filter node
+    // inputs are currently not pruned or eliminated via dead column elimination.
+    // TODO(todd): Investigate whether the shotgun filter node issue affects other
+    // query plans, i.e. filters before joins, and whether there is a more general
+    // approach to solving this (will still need the preceding project node for
+    // window functions preceded by filter nodes for correctness though)
+
+    if (!((always_add_project_if_first_project_is_window_expr &&
+           project_node_counter == 1) ||
+          filter_node || has_multi_fragment_scan_input)) {
+      continue;
+    }
+
+    InputCollector input_collector;
+    for (size_t i = 0; i < window_func_project_node->size(); i++) {
+      input_collector.visit(window_func_project_node->getExpr(i).get());
+    }
+    const InputSet& inputs = input_collector.result();
+
+    // Note: Technically not required since we are mapping old inputs to new input
+    // indices, but makes the re-mapping of inputs easier to follow.
+    std::vector<std::pair<const hdk::ir::Node*, unsigned>> sorted_inputs(inputs.begin(),
+                                                                         inputs.end());
+    std::sort(sorted_inputs.begin(),
+              sorted_inputs.end(),
+              [](const auto& a, const auto& b) { return a.second < b.second; });
+
+    hdk::ir::ExprPtrVector exprs;
+    std::vector<std::string> fields;
+    std::unordered_map<unsigned, unsigned> old_index_to_new_index;
+    for (auto& input : sorted_inputs) {
+      CHECK_EQ(input.first, prev_node.get());
+      auto res =
+          old_index_to_new_index.insert(std::make_pair(input.second, exprs.size()));
+      CHECK(res.second);
+      exprs.emplace_back(hdk::ir::makeExpr<hdk::ir::ColumnRef>(
+          getColumnType(input.first, input.second), input.first, input.second));
+      fields.emplace_back("");
+    }
+
+    auto new_project =
+        std::make_shared<hdk::ir::Project>(std::move(exprs), fields, prev_node);
+    node_list.insert(node_itr, new_project);
+    window_func_project_node->replaceInput(
+        prev_node, new_project, old_index_to_new_index);
+  }
+
+  // Any applied transformation always increases number of nodes.
+  if (node_list.size() != nodes.size()) {
+    nodes.assign(node_list.begin(), node_list.end());
+    dag.setNodes(std::move(nodes));
+  }
+}
+
 }  // namespace
 
 void canonizeQuery(QueryDag& dag) {
```

Comment:
I should have realized this before - but should this be `canonicalizeQuery`?

Reply:
Oops, sure it should.